hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/AbstractBatchDecoder.java

```diff
@@ -0,0 +1,442 @@
+                                                          @Override
+                                                          public ByteBuf cumulate(ByteBufAllocator alloc,
+                                                                                  ByteBuf cumulation,
+                                                                                  ByteBuf in) {
+                                                              ByteBuf buffer;
+                                                              if (cumulation.writerIndex() > cumulation
+                                                                  .maxCapacity()
+                                                                                             - in.readableBytes()
+                                                                  || cumulation.refCnt() > 1) {
+                                                                  // Expand cumulation (by replace it) when either there is not more room in the buffer
+                                                                  // or if the refCnt is greater then 1 which may happen when the user use slice().retain() or
+                                                                  // duplicate().retain().
+                                                                  //
+                                                                  // See:
+                                                                  // - https://github.com/netty/netty/issues/2327
+                                                                  // - https://github.com/netty/netty/issues/1764
+                                                                  buffer = expandCumulation(alloc,
+                                                                      cumulation,
+                                                                      in.readableBytes());
+                                                              } else {
+                                                                  buffer = cumulation;
+                                                              }
+                                                              buffer.writeBytes(in);
+                                                              in.release();
+                                                              return buffer;
+                                                          }
+                                                      };
+
+    /**
+     * Cumulate {@link ByteBuf}s by add them to a {@link CompositeByteBuf} and so do no memory copy whenever possible.
+     * Be aware that {@link CompositeByteBuf} use a more complex indexing implementation so depending on your use-case
+     * and the decoder implementation this may be slower then just use the {@link #MERGE_CUMULATOR}.
+     */
+    public static final Cumulator COMPOSITE_CUMULATOR = new Cumulator() {
+                                                          @Override
```

Comment:
Formatting is very strange shifted way to the right.

Reply:
@saintstack I reformat AbstractBatchDecoder class file.